x-pack/plugins/aiops/public/components/log_categorization/use_categorize_request.ts

```diff
@@ -0,0 +1,197 @@
+          )
+          .subscribe({
+            next: (result) => {
+              if (isCompleteResponse(result)) {
+                resolve(processCategoryResults(result, field));
+              } else if (isErrorResponse(result)) {
+                reject(result);
+              } else {
+                // partial results
+                // Ignore partial results for now.
+                // An issue with the search function means partial results are not being returned correctly.
+              }
+            },
+            error: (error) => {
+              if (error.name === 'AbortError') {
+                return resolve({ categories: [], sparkLinesPerCategory: {} });
+              }
+              reject(error);
+            },
+          });
+      });
+    },
+    [data.search]
+  );
+
+  const cancelRequest = useCallback(() => {
+    abortController.current.abort();
+    abortController.current = new AbortController();
+  }, []);
+
+  return { runCategorizeRequest, cancelRequest };
+}
+
+function createCategoryRequest(
+  index: string,
+  field: string,
+  timeField: string,
+  from: number | undefined,
+  to: number | undefined,
+  queryIn: QueryDslQueryContainer,
+  intervalMs?: number
+) {
+  const query = cloneDeep(queryIn);
+
+  if (query.bool === undefined) {
+    query.bool = {};
+  }
+  if (query.bool.must === undefined) {
+    query.bool.must = [];
+    if (query.match_all !== undefined) {
+      query.bool.must.push({ match_all: query.match_all });
+      delete query.match_all;
+    }
+  }
+  if (query.multi_match !== undefined) {
+    query.bool.should = {
+      multi_match: query.multi_match,
+    };
+    delete query.multi_match;
+  }
+
+  (query.bool.must as QueryDslQueryContainer[]).push({
+    range: {
+      [timeField]: {
+        gte: from,
+        lte: to,
+        format: 'epoch_millis',
+      },
+    },
+  });
+  return {
+    params: {
+      index,
+      size: 0,
+      body: {
+        query,
+        aggs: {
+          categories: {
+            categorize_text: {
+              field,
+              size: CATEGORY_LIMIT,
+            },
+            aggs: {
+              hit: {
+                top_hits: {
+                  size: EXAMPLE_LIMIT,
+                  sort: [timeField],
+                  _source: field,
+                },
+              },
+              ...(intervalMs
+                ? {
+                    sparkline: {
+                      date_histogram: {
+                        field: timeField,
+                        fixed_interval: `${intervalMs}ms`,
+                      },
+                    },
+                  }
+                : {}),
+            },
+          },
+        },
+      },
+    },
+  };
+}
+
+function processCategoryResults(result: CatResponse, field: string) {
+  const sparkLinesPerCategory: SparkLinesPerCategory = {};
+
+  const categories: Category[] = result.rawResponse.aggregations!.categories.buckets.map((b) => {
```

Comment:
To get rid of the `!` you could add a check before that line that it is set, something like:

```
    if (result.rawResponse.aggregations === undefined) {
      throw new Error('processCategoryResults failed, did not return aggregations.');
    }
```

We do similar checks in the explain log rate spikes code. We had cases with large datasets and clusters returning not an error but a differently formed response where it proofed worthy having that runtime check.

Reply:
Updated in https://github.com/elastic/kibana/pull/139005/commits/904f334a22cb9c3d26c316a179eeed9c5678090c